pkg/tbtc/dkg_submit.go

```diff
@@ -114,66 +127,77 @@ func (drs *dkgResultSubmitter) SubmitResult(
+				"aborting DKG result on-chain submission",
 			memberIndex,
 		)
 		return nil
 	}
 
-	submissionDelay := time.Duration(memberIndex-1) * dkgResultSubmissionDelayStep
+	blockCounter, err := drs.chain.BlockCounter()
+	if err != nil {
+		return err
+	}
+
+	// We can't determine a common block at which the publication starts.
+	// However, all we want here is to ensure the members does not submit
+	// in the same time. This can be achieved by simply using the index-based
+	// delay starting from the current block.
+	currentBlock, err := blockCounter.CurrentBlock()
+	if err != nil {
+		return fmt.Errorf("cannot get current block: [%v]", err)
+	}
+	delayBlocks := uint64(memberIndex-1) * dkgResultSubmissionDelayStepBlocks
+	submissionBlock := currentBlock + delayBlocks
 
 	drs.dkgLogger.Infof(
-		"[member:%v] waiting [%v] to submit",
+		"[member:%v] waiting for block [%v] to submit DKG result",
 		memberIndex,
-		submissionDelay,
+		submissionBlock,
 	)
 
-	submissionTimer := time.NewTimer(submissionDelay)
-	defer submissionTimer.Stop()
-
-	for {
-		select {
-		case <-submissionTimer.C:
-			// Member becomes eligible to submit the result. Result submission
-			// would trigger the sender side of the result submission event
-			// listener but also cause the receiver side (this select)
-			// termination that will result with a dangling goroutine blocked
-			// forever on the `onSubmittedResultChan` channel. This would
-			// cause a resource leak. In order to avoid that, we should
-			// unsubscribe from the result submission event listener before
-			// submitting the result.
-			subscription.Unsubscribe()
-
-			publicKeyBytes, err := result.GroupPublicKeyBytes()
-			if err != nil {
-				return fmt.Errorf("cannot get public key bytes [%w]", err)
-			}
-
-			drs.dkgLogger.Infof(
-				"[member:%v] submitting DKG result with public key [0x%x] and "+
-					"[%v] supporting member signatures",
-				memberIndex,
-				publicKeyBytes,
-				len(signatures),
-			)
-
-			return drs.chain.SubmitDKGResult(
-				memberIndex,
-				result,
-				signatures,
-			)
-		case blockNumber := <-resultSubmittedChan:
-			drs.dkgLogger.Infof(
-				"[member:%v] leaving; DKG result submitted by other member "+
-					"at block [%v]",
-				memberIndex,
-				blockNumber,
-			)
-			// A result has been submitted by other member. Leave without
-			// publishing the result.
-			return nil
-		case <-ctx.Done():
-			return fmt.Errorf("result publication timed out")
-		}
+	err = drs.waitForBlockFn(ctx, submissionBlock)
+	if err != nil {
+		return fmt.Errorf(
+			"error while waiting for DKG result submission block: [%v]",
+			err,
+		)
 	}
+
+	if ctx.Err() != nil {
+		// The context was cancelled by the upstream. Regardless of the cause,
+		// that means the DKG is no longer awaiting the result, and we can
+		// safely return.
+		drs.dkgLogger.Infof(
+			"[member:%v] DKG is no longer awaiting the result; "+
+				"aborting DKG result on-chain submission",
+			memberIndex,
+		)
```

Comment:
Is this always true? I think we do not know why the context got canceled, this could also be something else than the timeout 🤔 Isn't this assumption too much here?

Shouldn't we, for example:
- log here that we are aborting result submission because the parent context got cancelled
- when cancelling the parent context log that it's because of a timeout
?

Reply:
In this case, the cancellation of parent context means the result was submitted or a timeout was hit. In both cases, the DKG no longer awaits the result. Hence, in my opinion, that log makes sense and is more verbose than a mysterious "context got cancelled" that often causes confusion, even among keep-core developers.